### PR TITLE
[CALCITE-4041] Implement trait propagation for EnumerableCorrelate

### PR DESCRIPTION
**Pull request by rubenada**
Implement trait propagation for EnumerableCorrelate

### CODE REVIEW

**Comment by rubenada** [comment]
@hsyuan when you have some time, please take a look at the preliminary work on this PR (this is the first time that I touch trait propagation methods, so I am not very familiar with this subject).
To be done: unit tests

**Comment by amaliujia** [comment]
Indeed it looks good.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java [hunk 1]
```diff
@@ -81,6 +87,35 @@ public static EnumerableCorrelate create(
         traitSet, left, right, correlationId, requiredColumns, joinType);
   }
 
+  @Override public Pair<RelTraitSet, List<RelTraitSet>> passThroughTraits(
+      final RelTraitSet required) {
+    final RelCollation collation = required.getCollation();
+    if (collation == null || collation == RelCollations.EMPTY) {
+      return null;
+    }
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:95]: Not very familiar with `EnumerableCorrelate`, will it need to deal with RIGHT OUTER/FULL OUTER JOIN style case?

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:95]: Subquery, lateral query can only be inner, left outer correlate.
I don't know how to generate full/right outer correlate.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:95]: Correlate operator (base class) can only handle LEFT / INNER / SEMI / ANTI join. We do not need to consider RIGHT/ FULL in this case.

**Comment by rubenada** [comment]
Thanks for the feedback @hsyuan , @amaliujia . I have added some unit tests.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java [hunk 2]
```diff
@@ -81,6 +87,35 @@ public static EnumerableCorrelate create(
...
+    final RelTraitSet passThroughTraitSet = traitSet.replace(collation);
+    return Pair.of(passThroughTraitSet,
+        ImmutableList.of(
+            passThroughTraitSet,
+            passThroughTraitSet.replace(RelCollations.EMPTY)));
+  }
+
+  @Override public DeriveMode getDeriveMode() {
+    return DeriveMode.LEFT_FIRST;
+  }
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:117]: `deriveTraits` is still missing. If the left child can deliver some collation, the correlate should be able to deliver the same collation.

File: core/src/test/java/org/apache/calcite/test/TopDownOptTest.java [hunk 3]
```diff
@@ -132,6 +134,72 @@
...
+    final String sql = "select * from emp e\n"
+        + "join dept d on e.deptno=d.deptno\n"
+        + "order by e.ename";
+    Query.create(sql)
+        .addRule(JoinToCorrelateRule.INSTANCE)
+        .removeRule(EnumerableRules.ENUMERABLE_JOIN_RULE)
+        .removeRule(EnumerableRules.ENUMERABLE_MERGE_JOIN_RULE)
+        .removeRule(EnumerableRules.ENUMERABLE_SORT_RULE)
+        .check();
+  }
```
Comment (hsyuan) [inline core/src/test/java/org/apache/calcite/test/TopDownOptTest.java:148]: You can also use lateral to generate correlate if you would like to. The current tests are good enough, though.
e.g.
https://github.com/apache/calcite/blob/master/core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml#L6581

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java (hunk 2 above)
Comment (amaliujia) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:117]: Yes. Any collation can be derived from left input. Optimizer can decide later if derived collations are useful.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:131]: `deriveTraits` method added, not sure if the implementation is correct though, please take a look.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java [hunk 4]
```diff
@@ -81,6 +87,49 @@ public static EnumerableCorrelate create(
...
+    // should only derive traits (limited to collation for now) from left input.
+    assert childId == 0;
+
+    final RelCollation collation = childTraits.getCollation();
+    if (collation == null || collation == RelCollations.EMPTY) {
+      return null;
+    }
+
+    final RelTraitSet traits = traitSet.replace(collation);
+    return Pair.of(traits, ImmutableList.of(traits));
```
Comment (hsyuan) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java:126]: the list should contain right child traitset.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableCorrelate.java]: modified

File: core/src/test/java/org/apache/calcite/test/TopDownOptTest.java [hunk 5]
```diff
@@ -132,6 +134,72 @@
...
+
+  // Order by contains right field: sort cannot be pushed down.
+  @Test void testCorrelateInnerJoinNoDerive() {
+    final String sql = "select * from emp e\n"
+        + "join dept d on e.deptno=d.deptno\n"
+        + "order by e.ename, d.name";
```
Comment (amaliujia) [inline core/src/test/java/org/apache/calcite/test/TopDownOptTest.java:154]: Thanks for adding such test cases!

**Pull request closed without merging**